beacon_node/beacon_chain/src/validator_monitor.rs

```diff
@@ -491,6 +535,107 @@ impl<T: EthSpec> ValidatorMonitor<T> {
         }
     }
 
+    /// Process the unaggregated attestations generated by the service `attestation_simulator_service`
+    /// and check if the attestation qualifies for a reward matching the flags source/target/head
+    fn process_unaggregated_attestations(&mut self, state: &BeaconState<T>, spec: &ChainSpec) {
+        let current_slot = state.slot();
+
+        // Ensures that we process attestation when there have been skipped slots between blocks
+        let attested_slots: Vec<(Slot, Attestation<T>)> = self
+            .unaggregated_attestations
+            .iter()
+            .filter(|(attestation_slot, _)| {
+                **attestation_slot
+                    < current_slot - Slot::new(UNAGGREGATED_ATTESTATION_LAG_SLOTS as u64)
+            })
+            .map(|(slot, attestation)| (*slot, (*attestation).clone()))
+            .collect();
+
+        let unaggregated_attestations = &mut self.unaggregated_attestations;
+        for (slot, unaggregated_attestation) in attested_slots {
+            if unaggregated_attestations.remove_entry(&slot).is_some() {
```

Comment:
```suggestion
        // Ensures that we process attestation when there have been skipped slots between blocks
        let attested_slots: Vec<_> = self
            .unaggregated_attestations
            .iter()
            .map(|(attestation_slot, _)| *attestation_slot)
            .filter(|attestation_slot| {
                *attestation_slot
                    < current_slot - Slot::new(UNAGGREGATED_ATTESTATION_LAG_SLOTS as u64)
            })
            .collect();

        let unaggregated_attestations = &mut self.unaggregated_attestations;
        for slot in attested_slots {
            if let Some(unaggregated_attestation) = unaggregated_attestations.remove(&slot) {
```

It's very minor, but this would avoid us needing to clone the `unaggregated_attestation`.

Reply:
I am not sure I can do this without the borrow checker raising some concern.

We are using the result of `unaggregated_attestation.keys()` - which structure has for definition `HashMap<Slot, Attestation<T>>` - when prunning the hashmap a little below `unaggregated_attestations.remove(&slot)`.

The borrow checker doesn't like it as `keys()` asks for an immutable reference of `unaggregated_attestations` and we need a mutable reference for prunning. I have added your recommendation but I have also had to still copy the retained slots.

Let me know what you think.

Reply:
Good idea of using `.keys()`, that's better than my suggestion!

Copying the `Slot` is totally fine, it's only 8 bytes. A reference (on 64-bit platforms) is the same size as a `Slot` so using a `Vec<Slot>` is going to use the same amount of memory as a `Vec<&Slot>`, whilst being more direct.